README.md
# Guardrails Memory Debugger

## Background

Guardrails was written years ago to debug a highly parallel distributed OLAP
database.  The database had a custom runtime that managed many user-mode fibers
with makecontext, swapcontext and getcontext.  Most of the
test/verification/automation infrastructure assumed clusters consisting of
many-core machines.  This resulted in very poor valgrind performance due to
thread serialization. Older ASAN implementations had difficulty dealing with
the swapcontext calls.  Thus guardrails was born.

These days you should probably start with tools like ASAN, valgrind, radare,
etc.  However the relative simplicity of the guardrails approach allows a lot
of flexibility in modifying the tool itself for specific debugging tasks. Note
that guardrails performance will be suboptimal under very heavy heap churn of
small allocations. In the case of the OLAP database, this was of minimal impact
because the heap was mostly used for the control path (very complex but
relatively lightly loaded) and not the data path, which was the performance
critical path but largely relied on larger allocations via a custom slab
allocator.

## Introduction

GuardRails is a MMU-based memory debugging tool.  The objective is to support
similar functionality to valgrind but with much higher performance in highly
threaded environments.  GuardRails is a full replacement for the libc memory
allocator.  Calls such as malloc, memalign, free, realloc, etc. are all handled
by GuardRails.  This is achieved using the LD_PRELOAD environment variable to
force the loader to overlay the GuradRails shared library over libc.

Valgrind is a great debugging tool but is far too slow to run massively
parallel tests, especially on many-core machines. The Valgrind VM alone running
memcheck exhibits a [50x performance degradation over
native](http://valgrind.org/docs/manual/manual-core.html#manual-core.whatdoes).
The problem is further compounded by the fact that Valgrind [serializes
threads](http://valgrind.org/docs/manual/manual-core.html#manual-core.pthreads),
limiting execution to a single core regardless of machine size.  These two
factors together can result in a <b>performance degradation factor of over
1,000x</b> for many-core machines.  Aside from the performance implications,
Valgrind fully virtualizes program execution, resulting in other undesired
side-effects related to shared memory, signal handling, etc.

## Guardrails tools
* `grrun.sh` runs executables under guardrails
* `test` a very simple test program exercising some guardrails functionality (see `./test -h`)
* `grgdb.py` GDB debugging macros
* `grdump.py` leak dumping utility
* `libguardrails.so.0.0` guardrails shared library; provides dumb best-fit
  logarithmic allocator along with tracking and protection mechanisms

# GuardRails Debugging Features

GuardRails uses the hardware MMU to detect memory errors by modifying the page
protection bits as memory is allocated and freed.  Before actually being
recycled freed memory is added to a delay list during which time the pages are
fully read/write protected.  This ensures that any use-after-free causes a
segmentation fault.  Heap header protection is achieved by appending a
protected guard page adjacent to all memory allocations.  Any corruption of
this header will cause a segfault indicating who did the corruption.

Current features include:

* Protection of the heap metadata.  This catches the infamous "__libc_message" errors from the libc heap allocator.
* Buffer overrun detection with allocation and free tracking.
* Context aware leak tracking with full allocator backtrace.  Both size and frequency of leaks are tracked and reported.
* Double free detection with allocation and free tracking.
* Use-after-free detection with allocation and free tracking.
* Memory poisoning and verification.
* Delayed free lists.
* Various OOM handling options.
* High-watermark tracking with dump intervals.
* Debugging of transient hidden runtime leaks (leaks that are freed via destructors on program exit).
* Various stats such as allocation histograms, rates, etc.

Future features:

* Better test infra
* Improved use-after-free performance.
* Buffer underrun detection.
* Thread-local memory
* Lockless pools

# Building Guardrails

A full build can be done with:

```console
make deps
make
```

This will result in a shared library `libguardrails.so.0.0` build product and a test program `test`.

# Running Under Guardrails

You'll may need to increase the `max_map_count` proc node to prevent allocation failures with more complex allocation patterns:

`echo 10000000 | sudo tee /proc/sys/vm/max_map_count`

and

`echo 0 | sudo tee /proc/sys/kernel/randomize_va_space`

Then use the wrapper script to run the program:

`grrun.sh <program>`

For example:

`grrun.sh ./test -o`

to run the simple buffer overrun test (expected to segfault).

No need to recompile the program; guardrails can run against any arbitrary binary.

# GuardRails Debugging Techniques

GuardRails provides several facilities for both detecting and also rapidly
fixing memory related bugs. In general, <b><u>when such bugs are detected the program
will immediately segfault</u></b>.  Debugging can then commence via an attached gdb
session or core file (limited functionality at the moment) using the supplied
GDB macros.

The following techniques are useful when running under GuardRails.

## Memory Debugging

In general GuardRails works by causing the process to segfault the moment
corruption occurs.  Examining the backtrace from the core will show exactly
what caused the corruption.  Guardrails also adds significant additional
context, including the original allocator and original free'er of the memory in
question.

### GDB Macros

Several macros are provided to facilitate GuardRails debugging.  These can be loaded into GDB as follows:

`(gdb) source grgdb.py`

See "help <macro>" for up-to-date help and examples.

Current macros include:

* <b>gr-dump-in-flight</b>: Dump allocator metadata. This is mostly for dumping
  in-flight allocations to a file for subsequent post-processing with
  grdump.py.  This is NOT recommended for interactive debugging.

* <b>gr-find-header</b>: Given an arbitrary address, find the GuardRails header
  without depending on the delay list.  The output of this command can be used
  as input to gr-print-addr-info.

* <b>gr-find-delay-list</b>: Finds an address on the delayed free list
  (requires -d option when running GuardRails).  Any address within an
  allocated range can be supplied here.  This is useful whenever a segfault
  occurs, as the faulting address is likely to be on this list, indicating a
  corruption, use-after-free or double free.

* <b>gr-heap-meta-corruption</b>: Try to determine if a faulting address would
  cause heap corruption. This command indicates if accessing a given address
  would cause corruption by determining if the access falls on a guard page.

* <b>gr-print-addr-info</b>: Prints information about an address.  This
  includes the GuardRails header along with the backtraces of the allocator of
  the memory and the freer of the memory.  In general this information should
  result in an almost immediate fix.  Note that the address MUST be either the
  original address returned by the allocator (eg malloc), or the GuardRails
  header address.  If the address is some other pointer, use gr-find-header to
  get the associated header address.

* <b>gr-print-segv</b>: Print segfault address.

#### GDB Macros Example Session

What follows is an example session debugging a corruption using GuardRails GDB macros.  It assumes GuardRails was run with the the following command line in `grargs.txt`:

`-dvt30 -T30 -p 175`

GDB session workflow (of course all the standard GDB debugging facilities are also available but omitted here for brevity):

```
Program terminated with signal SIGSEGV, Segmentation fault.
#0  0x00007fc6af5b4818 in json_decref (json=0x7fc6aa5f2fd0) at jansson.h:111
111         if(json && json->refcount != (size_t)-1 && --json->refcount == 0)
(gdb) source grgdb.py

# The actual address that caused the segfault
(gdb) gr-print-segv
$1 = (void *) 0x7fc6aa5f2fd8

# Find the GuardRails header associated with that address
(gdb) gr-find-header 0x7fc6aa5f2fd8
Found valid header at: 0x7fc6aa5f2000

# Dump the header and traces associated with that address
(gdb) gr-print-addr-info 0x7fc6aa5f2000
Address 0x7fc6aa5f2000 is a header
Address 0x7fc6aa5f2000 is free
Header:
{magic = 14800349170826394398, binNum = 12, slotNum = 0, usrData = 0x7fc6aa5f2fc0, usrDataSize = 64, next = 0x0, prev = 0x0, allocBt = 0x7fc6aa5f2038}

# Backtrace that allocated the memory:
================ Allocation Trace: ================
Line 325 of "GuardRails.c" starts at address 0x7fc6af7c0f35 <memalignInt+357>
Line 299 of "../src/lib/libutil/MemTrack.cpp" starts at address 0xa9363f <_memAlloc(unsigned long, char const*, char const*)+127>
Line 996 of "../src/lib/libutil/MemTrack.cpp" starts at address 0xa960d9 <memAllocJson(unsigned long)+25>
Line 341 of "value.c" starts at address 0x7fc6af5b91fb <json_array+11>
Line 770 of "load.c" starts at address 0x7fc6af5b6185 <parse_value+741>
Line 734 of "load.c" starts at address 0x7fc6af5b609a <parse_value+506>
Line 734 of "load.c" starts at address 0x7fc6af5b609a <parse_value+506>
Line 779 of "load.c" starts at address 0x7fc6af5b61be <parse_value+798>
Line 898 of "load.c" starts at address 0x7fc6af5b6426 <parse_json+70>
Line 959 of "load.c" starts at address 0x7fc6af5b65c3 <json_loads+227>
Line 499 of "../src/lib/libqueryparser/QueryParser.cpp" starts at address 0x7582b6 <QueryParser::jsonParse(char const*, Dag*)+102>
Line 654 of "../src/lib/libqueryparser/QueryParser.cpp" starts at address 0x758d00 <QueryParser::parse(char const*, Dag**, unsigned long*)+240>
Line 2794 of "../src/lib/libdag/Retina.cpp" starts at address 0x7f2f06 <DagLib::importRetina(ApiImportRetinaInput*, ApiOutput**, unsigned long*, bool)+1110>
Line 37 of "../src/lib/libapis/ApiHandlerImportRetina.cpp" starts at address 0x7d1a3c <ApiHandlerImportRetina::run(ApiOutput**, unsigned long*)+140>
Line 73 of "../src/lib/libapis/ApisRecvObject.cpp" starts at address 0x79630f <ApisRecvObject::run()+639>
Line 1113 of "../src/lib/libapis/ApisRecv.cpp" starts at address 0x795e4b <doWorkForApiImmediate(void*)+363>
Line 65 of "../src/lib/libruntime/DedicatedThread.cpp" starts at address 0x9b55d8 <DedicatedThread::threadEntryPoint()+120>
Line 55 of "../src/lib/libruntime/Thread.cpp" starts at address 0x9bc911 <Thread::threadEntryPointWrapper(void*)+113>
Line 312 of "pthread_create.c" starts at address 0x7fc6ae722173 <start_thread+179>
Line 113 of "../sysdeps/unix/sysv/linux/x86_64/clone.S" starts at address 0x7fc6acbdc03d <clone+109>

# Backtrace that freed the memory:
================ Free Trace: ================
Line 755 of "../src/lib/libutil/MemTrack.cpp" starts at address 0xa93e99 <memFree(void*)+313>
Line 1002 of "../src/lib/libutil/MemTrack.cpp" starts at address 0xa96139 <memFreeJson(void*)+25>
Line 240 of "../src/lib/libqueryparser/QpMap.cpp" starts at address 0x765c80 <QpMap::parseJson(json_t*, json_error_t*, WorkItem**)+992>
Line 567 of "../src/lib/libqueryparser/QueryParser.cpp" starts at address 0x758718 <QueryParser::jsonParse(char const*, Dag*)+1224>
Line 654 of "../src/lib/libqueryparser/QueryParser.cpp" starts at address 0x758d00 <QueryParser::parse(char const*, Dag**, unsigned long*)+240>
Line 2794 of "../src/lib/libdag/Retina.cpp" starts at address 0x7f2f06 <DagLib::importRetina(ApiImportRetinaInput*, ApiOutput**, unsigned long*, bool)+1110>
Line 37 of "../src/lib/libapis/ApiHandlerImportRetina.cpp" starts at address 0x7d1a3c <ApiHandlerImportRetina::run(ApiOutput**, unsigned long*)+140>
Line 73 of "../src/lib/libapis/ApisRecvObject.cpp" starts at address 0x79630f <ApisRecvObject::run()+639>
Line 1113 of "../src/lib/libapis/ApisRecv.cpp" starts at address 0x795e4b <doWorkForApiImmediate(void*)+363>
Line 65 of "../src/lib/libruntime/DedicatedThread.cpp" starts at address 0x9b55d8 <DedicatedThread::threadEntryPoint()+120>
Line 55 of "../src/lib/libruntime/Thread.cpp" starts at address 0x9bc911 <Thread::threadEntryPointWrapper(void*)+113>
Line 312 of "pthread_create.c" starts at address 0x7fc6ae722173 <start_thread+179>
Line 113 of "../sysdeps/unix/sysv/linux/x86_64/clone.S" starts at address 0x7fc6acbdc03d <clone+109>

# Backtrace of the code that tried to use the memory after it was freed:
(gdb) bt
#0  0x00007fc6af5b4818 in json_decref (json=0x7fc6aa5f2fd0) at jansson.h:111
#1  hashtable_do_clear (hashtable=hashtable@entry=0x7fc6aa604fc0) at hashtable.c:148
#2  0x00007fc6af5b48d9 in hashtable_close (hashtable=0x7fc6aa604fc0) at hashtable.c:215
#3  0x00007fc6af5b97a1 in json_delete_object (object=0x7fc6aa604fb0) at value.c:77
#4  json_delete (json=0x7fc6aa604fb0) at value.c:936
#5  0x00007fc6af5b4834 in json_decref (json=<optimized out>) at jansson.h:112
#6  hashtable_do_clear (hashtable=hashtable@entry=0x7fc6aa60efc0) at hashtable.c:148
#7  0x00007fc6af5b48d9 in hashtable_close (hashtable=0x7fc6aa60efc0) at hashtable.c:215
#8  0x00007fc6af5b97a1 in json_delete_object (object=0x7fc6aa60efb0) at value.c:77
#9  json_delete (json=0x7fc6aa60efb0) at value.c:936
#10 0x00007fc6af5b9815 in json_decref (json=<optimized out>) at jansson.h:112
#11 json_delete_array (array=<optimized out>) at value.c:364
#12 json_delete (json=0x7fc6aa54efd0) at value.c:939
#13 0x0000000000758b5d in json_decref (json=0x7fc6aa54efd0) at /usr/include/jansson.h:112
#14 QueryParser::jsonParse (this=0x7fc66732b790, query=0x7fc67696ea10 "[{\"op"..., queryGraph=0x7fc491655a80) at ../src/lib/libqueryparser/QueryParser.cpp:619
#15 0x0000000000758d11 in QueryParser::parse (this=0x7fc66732b790, query=0x7fc67696ea10 "[{\"op"..., queryGraphOut=0x7fc476ca95b8, numQueryGraphNodesOut=0x7fc476ca95b0) at ../src/lib/libqueryparser/QueryParser.cpp:654
#16 0x00000000007f2f26 in DagLib::importRetina (this=0x7fc66a42af60, importRetinaInput=0x7fc6a9a28328, importRetinaOutputOut=0x7fc491651030, importRetinaOutputSizeOut=0x7fc491651038, persist=true) at ../src/lib/libdag/Retina.cpp:2794
#17 0x00000000007d1a44 in ApiHandlerImportRetina::run (this=0x7fc6a9a1ef80, output=0x7fc491651030, outputSize=0x7fc491651038) at ../src/lib/libapis/ApiHandlerImportRetina.cpp:37
#18 0x0000000000796345 in ApisRecvObject::run (this=0x7fc491651000) at ../src/lib/libapis/ApisRecvObject.cpp:73
#19 0x0000000000795e4b in apisRecvObjectImmediateHandler (apisRecvObject=0x7fc491651000) at ../src/lib/libapis/ApisRecv.cpp:1112
#20 doWorkForApiImmediate (args=0x4) at ../src/lib/libapis/ApisRecv.cpp:1138
#21 0x00000000009b55ec in DedicatedThread::threadEntryPoint (this=0x7fc67382edd8) at ../src/lib/libruntime/DedicatedThread.cpp:65
#22 0x00000000009bc91e in Thread::threadEntryPointWrapper (arg=0x7fc67382edd8) at ../src/lib/libruntime/Thread.cpp:55
#23 0x00007fc6ae722184 in start_thread (arg=0x7fc476cb0700) at pthread_create.c:312
#24 0x00007fc6acbdc03d in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:111
```

## SIGUSR2

Issuing SIGUSR2 to &lt;program_name&gt; running under GuardRails:

`pkill -SIGUSR2 <program_name>`

will cause the current GuardRails state to be dumped to stdout and also cause
GuardRails to write out the grtrack-&lt;tid&gt;.txt leak tracking file
(assuming the -t and/or -T option was passed).  Included in the state is the
arguments from grargs.txt, which can be used to verify that GuardRails is
running as intended.  If the -v option was passed, a per-memory-bin stats will
also be dumped.

Note that depending on the allocator state, it can take several minutes to dump
the tracker file.  To determine when it's done, you can do something like

`watch -n1 ls -al grtracker-*.txt`

to see when it stops growing.  lsof can also be used.  Further, the file is
also done dumping when the allocator state summary is finished dumping to
node.out (can tail -f this).

Note that grdump.py can also handle partial dumps, which might be sufficient
depending on the circumstances.

## Leak Tracking

Leak tracking must be enabled by adding the "-t <frames>" and/or "-T <frames>"
option to the grargs.txt file.  There are some deep stacks in the protobuf code
paths.  In general a max frame count of 30 is sufficient to get enough context
in all such cases.

When tracking leaks, GuardRails will dump an output CSV file called
"grtrack.txt" to the current working directory.  This file is dumped on exit
and also on SIGUSR2.  It can be post-processed by grdump.py.  This script
groups and counts all the leak contexts, generates a symbolic backtrace for
each context, and sorts them by either the total number of bytes leaked or by
the frequency of the leak.

If instead SIGUSR2 is used to dump grtrack.txt, it will contain not only leaked
memory but also any in-use memory at the time SIGUSR2 was issued.

### Leak Tracking Examples

To show the top 100 leaks by frequency:

<pre>
./grdump.py -f grtrack-&lt;tid&gt;.txt -b &lt;program_name&gt; -c -t 100 |less
Leaked total of 4,222,380 bytes across 62,116 leaks from 1,639 contexts
================================ Context      0 / 1,639 ================================
Leaked 171,008 bytes across 5,344 allocations of 32 bytes each:
#0  0x00007f457e430de9: ?? ??:0
#1  0x000000000091fe24: string_create at value.c:655
#2  0x000000000091ce60: parse_value at load.c:830
#3  0x000000000091cf6a: parse_object at load.c:734
#4  0x000000000091cf6a: parse_object at load.c:734
#5  0x000000000091d08e: parse_array at load.c:779
#6  0x000000000091d2f6: parse_json at load.c:898
#7  0x000000000091d46b: json_loads at load.c:959
#8  0x000000000069996f: QueryParser::jsonParse at QueryParser.cpp:499
#9  0x0000000000699de8: QueryParser::parse at QueryParser.cpp:648
#10 0x00000000006f12cb: DagLib::copyRetinaToNewDag at Retina.cpp:5947
#11 0x00000000006e02ea: Dag::copyDagToNewDag at Dag.cpp:2105
#12 0x00000000006e16f3: Dag::cloneDagLocal at Dag.cpp:2800
#13 0x00000000006e291a: Dag::cloneDagGlobalInt [256], unsigned long, char  [256], Dag::CloneFlags) at Dag.cpp:3676
#14 0x00000000006e2551: Dag::cloneDagGlobal [256], unsigned int, char  [256], Dag::CloneFlags) at Dag.cpp:3488
#15 0x00000000006ec54b: DagLib::makeRetina at Retina.cpp:2980
#16 0x00000000006d453b: ApiHandlerMakeRetina::run at ApiHandlerMakeRetina.cpp:50
#17 0x000000000075b569: processWorkItem at OperatorsFuncTest.cpp:413
#18 0x000000000076128e: workerRetina at OperatorsFuncTest.cpp:4359
#19 0x000000000075e6b2: workerMain at OperatorsFuncTest.cpp:4898
#20 0x00000000007bd835: DedicatedThread::threadEntryPoint at DedicatedThread.cpp:65
#21 0x00000000007bfe0f: Thread::threadEntryPointWrapper at ??:?
#22 0x00007f457d598e25: ?? ??:0
#23 0x00007f457bc5f34d: ?? ??:0
================================ Context      1 / 1,639 ================================
Leaked 84,096 bytes across 2,628 allocations of 32 bytes each:

remaining top 99 leaks elided...
</pre>

To show the top 100 by total leaked memory:

<pre>
./grdump.py -f grtrack-&lt;tid&gt;.txt -b &lt;program_name&gt; -t 100 |less
Leaked total of 4,222,380 bytes across 62,116 leaks from 1,639 contexts
================================ Context      0 / 1,639 ================================
Leaked 902,472 bytes across 1,213 allocations of 744 bytes each:
#0  0x00007f457e430de9: ?? ??:0
#1  0x00007f457e18524d: ?? ??:0
#2  0x000000000078b7c7: resultSetMakeLocal at LocalResultSet.cpp:147
#3  0x000000000078b401: usrNodeMsg2pcMakeResultSet at UsrNode.cpp:369
#4  0x00000000006910fa: MsgMgr::doTwoPcWorkForMsg at Message.cpp:3387
#5  0x00000000007bfcf5: SchedObject::getSchedulable at SchedObject.h:49
#6  0x00000000007bd551: memBarrier at System.h:65
#7  0x00007f457bbadd40: ?? ??:0

changes.txt
commit 7d3ca8726e9c1a80292ca7655bce628caa4f1fdd (HEAD -> master, origin/master, origin/HEAD)
Date:   Wed Jan 20 09:51:12 2021 -0800

    Move code to Apache2 License
    
    Copyright 2021, Xcalar Inc.
    
    Licensed under the Apache License, Version 2.0 (the "License");
    you may not use this file except in compliance with the License.
    You may obtain a copy of the License at
    
        http://www.apache.org/licenses/LICENSE-2.0
    
    Unless required by applicable law or agreed to in writing, software
    distributed under the License is distributed on an "AS IS" BASIS,
    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
    See the License for the specific language governing permissions and
    limitations under the License.
    
    Change-Id: I8471e1284e1e6cce95a771e7bcf83f1517b7ddbe


grgdb.py
#
# Copyright 2021, Xcalar Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

# Author: Eric D. Cohen

# GDB commands for use with GuardRails.
# (gdb) source grgdb.py

import gdb
import gdb.printing
import re

PAGE_SIZE = 4096
NUM_GUARD_PAGES = 1
GUARD_SIZE = NUM_GUARD_PAGES * PAGE_SIZE
MAGIC_FREE  = 0xcd656727bedabb1e
MAGIC_INUSE = 0x4ef9e433f005ba11
MAGIC_GUARD = 0xfd44ba54deadbabe

def checkGr():
    try:
        gdb.parse_and_eval("grArgs")
    except gdb.error as e:
        msg = "FATAL GUARDRAILS ERROR: Program not run under guardrails"
        print(msg)
        sys.stderr.write(msg + "\n")
        raise

class GRFindDelayList (gdb.Command):
    """
        Search Guard Rails delayed free list for a given address.

        (gdb) gr-find-delay-list <delayListAddress> <addressToFind>

        Example:
        (gdb) gr-find-delay-list &memSlots[0].delay 0x7f607e24dbf0
    """

    def __init__ (self):
        super (GRFindDelayList, self).__init__ ("gr-find-delay-list", gdb.COMMAND_USER)

    def delayListCount(self, head, tail, maxDelayCt):
        if head > tail:
            return head - tail
        else:
            return maxDelayCt - (tail - head)

    def invokeHelper (self, arg, from_tty):
        argv = gdb.string_to_argv(arg)
        if len(argv) != 2:
            print("Malformed arguments; see help")
            return

        dlist = gdb.parse_and_eval('(MemFreeDelay *) ' + argv[0])
        addrToFind = int(argv[1], 16)
        dlistElms = dlist['elms']
        maxDelayCt = int(gdb.parse_and_eval('sizeof(((MemFreeDelay *)0x0)->elms)/sizeof(((MemFreeDelay *)0x0)->elms[0])'))
        head = dlist['head']
        tail = dlist['tail']
        cursor = head
        delayCount = self.delayListCount(head, tail, maxDelayCt)
        ct = 0
        while ct < delayCount:
            if not ct % 1000:
                print("Searching %7d / %d elements" % (ct, delayCount))
            elmPtr = dlistElms[cursor]
            elm = elmPtr.dereference()

            magicStr = str(elm['magic'])
            if elm['magic'] != MAGIC_FREE:
                print("Delayed free list contains invalid header magic 0x%x" % magicStr)

            elmStartAddr = int(elmPtr)

            # Consider the full allocation (not just the amount allocated for the
            # user), as the errant pointer might land anywhere in that range.
            elmEndAddr = elmStartAddr + (1 << int(elm['binNum'])) + GUARD_SIZE - 1

            if elmStartAddr <= addrToFind <= elmEndAddr:
                print("Found address 0x%x on delayed free list at index %d, header 0x%x :" % (addrToFind, cursor, elmPtr))
                # print("%x <= %x <= %x" % (elmStartAddr, addrToFind, elmEndAddr))
                print(elm)
                # An element should only appear on the list once, so return
                return

            # Search the list from the head (most recently added) backwards as
            # the element of interest is most likely recent (and search can be
            # somewhat slow)
            if cursor == 0:
                cursor = maxDelayCt
            cursor -= 1

            ct += 1

        print("Address 0x%x not found" % (addrToFind))


    def invoke(self, arg, from_tty):
        try:
            checkGr()
            self.invokeHelper(arg, from_tty)
        except Exception as e:
            print(str(e))
            traceback.print_exc()

class GRPrintAddrInfo (gdb.Command):
    """
        Try to print GuardRails info about an address.

        The address must be either the beginning of a header or beginning of a
        user allocation.  If the address is a random memory address not known
        to point to a header or beginning allocation, first try to find the
        associated header address using gr-find-header.

        If GuardRails was run with -t and/or -T options, traces will also be
        dumped for allocations/frees.

        (gdb) gr-print-addr-info <address>

        Example:
        (gdb) gr-print-addr-info 0x7fc6aa5f2000
    """

    def __init__ (self):
        super (GRPrintAddrInfo, self).__init__ ("gr-print-addr-info", gdb.COMMAND_USER)

    def isValid(self, magic):
        return magic == MAGIC_INUSE or magic == MAGIC_FREE

    def dumpSymTrace(self, trace, offset, maxFrames):
        if not maxFrames:
            print("Memory tracking not enabled; to enable rerun GuardRails with -t/-T options")
            return

        for i in range(offset, maxFrames + offset):
            addr = str(trace[i]).split(' ')[0]

            addrInt = int(addr, 0)
            if not addrInt:
                continue
            sym = str(gdb.execute('info line *' + str(addrInt), to_string=True))
            sym = re.sub(r' and ends at .*', r'', sym)
            print(sym[0:-1])

    def invokeHelper(self, arg, from_tty):
        argv = gdb.string_to_argv(arg)
        if len(argv) != 1:
            print("Malformed arguments; see help")
            return

        hdrPtr = gdb.parse_and_eval('(ElmHdr *) ' + argv[0])
        hdr = hdrPtr.dereference()
        grArgs = gdb.parse_and_eval("grArgs")
        maxAllocFrames = int(grArgs['maxTrackFrames'])
        maxFreeFrames = int(grArgs['maxTrackFreeFrames'])

        if self.isValid(hdr['magic']):
            print("Address %s is a header" % argv[0])
        else:
            hdrPtr = gdb.parse_and_eval('*(ElmHdr **)((char *)' + argv[0] + ' - sizeof(void *))')
            hdr = hdrPtr.dereference()
            if self.isValid(hdr['magic']):
                print("Address %s is a user address" % argv[0])
            else:
                print("Address %s doesn't look valid" % argv[0])
                return

        if hdr['magic'] == MAGIC_INUSE:
            print("Address %s is in-use" % argv[0])
        elif hdr['magic'] == MAGIC_FREE:
            print("Address %s is free" % argv[0])

        print("Header:")
        print(hdr)

        trace = hdr['allocBt']

        print("================ Allocation Trace: ================")
        self.dumpSymTrace(trace, 0, maxAllocFrames)
        print("================ Free Trace: ================")
        self.dumpSymTrace(trace, maxAllocFrames + 1, maxFreeFrames)

    def invoke(self, arg, from_tty):
        try:
            checkGr()
            self.invokeHelper(arg, from_tty)
        except Exception as e:
            print(str(e))
            traceback.print_exc()

class GRFindHeader(gdb.Command):
    """
        Try to find the GuardRails header address for an arbitrary address

        Use this to find the header address associated with an arbitrary memory
        address.  The output address of this command can be used with
        gr-print-addr-info.

        (gdb) gr-find-header <address>

        Example:
        (gdb) gr-find-header 0x7fc6aa5f2327
    """

    def __init__ (self):
        super (GRFindHeader, self).__init__ ("gr-find-header", gdb.COMMAND_USER)

    def isValid(self, magic):
        return magic == MAGIC_INUSE or magic == MAGIC_FREE

    def invokeHelper(self, arg, from_tty):
        argv = gdb.string_to_argv(arg)
        if len(argv) != 1 and (len(argv) == 2 and argv[1] != 'quiet') or len(argv) > 2:
            print("Malformed arguments; see help")
            return

        addr = int(argv[0], 16)
        quiet = False
        if len(argv) == 2:
            quiet = argv[1] == 'quiet'
        mask = 0xffffffffffffffff - PAGE_SIZE + 1
        headerStart = addr & mask

        currHeader = headerStart
        while (True):
            hdrPtr = gdb.parse_and_eval('((ElmHdr *) ' + str(currHeader) + ')')
            hdr = hdrPtr.dereference()
            if self.isValid(hdr['magic']):
                # XXX: Add header sanity checks here
                if quiet:
                    print("0x%x" % currHeader)
                else:
                    print("Found valid header at: 0x%x" % currHeader)
                break

            assert(currHeader > PAGE_SIZE)
            currHeader -= PAGE_SIZE

    def invoke(self, arg, from_tty):
        try:
            checkGr()
            self.invokeHelper(arg, from_tty)
        except Exception as e:
            print(str(e))
            traceback.print_exc()

class GRHeapMetaCorruption(gdb.Command):
    """
        Try to determine if a faulting address would cause heap corruption.

        This command indicates if accessing a given address would cause
        corruption by determining if the access falls on a guard page.

        (gdb) gr-heap-meta-corruption <address>

        Example:
        (gdb) gr-heap-meta-corruption 0x7fc6aa5f2327
    """

    def __init__ (self):
        super (GRHeapMetaCorruption, self).__init__ ("gr-heap-meta-corruption", gdb.COMMAND_USER)

    def invokeHelper(self, arg, from_tty):
        argv = gdb.string_to_argv(arg)
        if len(argv) != 1:
            print("Malformed arguments; see help")
            return

        addr = int(argv[0], 16)
        currHeader = gdb.execute("gr-find-header 0x%x quiet" % (addr), to_string=True)
        hdrPtr = gdb.parse_and_eval('((ElmHdr *) ' + str(currHeader) + ')')
        hdr = hdrPtr.dereference()
        hdrPtrAddr = int(hdrPtr)
        guardStartAddr = hdrPtrAddr + (1 << hdr['binNum'])
        guardEndAddr = guardStartAddr + GUARD_SIZE - 1

        verifyGuard = gdb.parse_and_eval("*((uint64_t *) 0x%x)" % guardStartAddr)
        assert(verifyGuard == MAGIC_GUARD)
        if guardStartAddr <= addr <= guardEndAddr:
            print("CORRUPTION: Access of 0x%x appears to corrupt the heap metadata and/or overrun the buffer" % addr)
        else:
            print("Access of 0x%x doesn't appear to be heap metadata corruption or overrun" % addr)

    def invoke(self, arg, from_tty):
        try:
            checkGr()
            self.invokeHelper(arg, from_tty)
        except Exception as e:
            print(str(e))
            traceback.print_exc()

class GRPrintSegv(gdb.Command):
    """
        Print address of access that caused the current segfault.

        (gdb) gr-print-segv

        Example:
        (gdb) gr-print-segv
    """

    def __init__ (self):
        super (GRPrintSegv, self).__init__ ("gr-print-segv", gdb.COMMAND_USER)

    def invokeHelper(self, arg, from_tty):
        argv = gdb.string_to_argv(arg)
        if len(argv) != 0:
            print("Malformed arguments; see help")
            return

        sigInfo = gdb.parse_and_eval("$_siginfo")
        if sigInfo['si_signo'] == 7 or sigInfo['si_signo'] == 11:
            print("Memory fault at: " + str(sigInfo['_sifields']['_sigfault']['si_addr']))
        else:
            print("Unknown fault, signal %d" % sigInfo['si_signo'])

    def invoke(self, arg, from_tty):
        try:
            self.invokeHelper(arg, from_tty)
        except Exception as e:
            print(str(e))
            traceback.print_exc()

class GRDumpInFlight(gdb.Command):
    """
        Dump allocator metadata.

        WARNING: This is mostly for dumping in-flight allocations to a file
        for subsequent post-processing with grdump.py.  This is NOT recommended
        for interactive debugging.

        (gdb) gr-dump-in-flight

        Example:
        (gdb) gr-dump-in-flight
    """

    def __init__ (self):
        super (GRDumpInFlight, self).__init__ ("gr-dump-in-flight", gdb.COMMAND_USER)

    def getTraceCsv(self, trace, offset, maxFrames):
        if not maxFrames:
            print("Memory tracking not enabled; to enable rerun GuardRails with -t/-T options")
            return

        ret = ""
        for i in range(offset, maxFrames + offset):
            addrInt = int(trace[i])
            if addrInt:
                ret += "0x%x," % addrInt
            else:
                ret += ","

        return ret

    def invokeHelper(self, arg, from_tty):
        argv = gdb.string_to_argv(arg)
        if len(argv) != 0:
            print("Malformed arguments; see help")
            return

        grArgs = gdb.parse_and_eval("grArgs")
        numSlots = int(grArgs['numSlots'])
        maxAllocFrames = int(grArgs['maxTrackFrames'])
        maxFreeFrames = int(grArgs['maxTrackFreeFrames'])
        memSlots = gdb.parse_and_eval("memSlots")

        print("===== START TRACES =====")
        for slotNum in range(0, numSlots):
            memBins = memSlots[slotNum]['memBins']
            numBins = int(gdb.parse_and_eval('sizeof(((MemSlot *)0x0)->memBins)/sizeof(((MemSlot *)0x0)->memBins)[0]'))
            for binNum in range(0, numBins):
                memBin = memBins[binNum]
                hdr = memBin['headInUse']
                while hdr != 0x0:
                    csvTrace = self.getTraceCsv(hdr['allocBt'], 0, maxAllocFrames)
                    elmSize = hdr['usrDataSize']
                    print(str(elmSize) + "," + csvTrace)
                    hdr = hdr['next']

    def invoke(self, arg, from_tty):
        try:
            checkGr()
            self.invokeHelper(arg, from_tty)
        except Exception as e:
            print(str(e))
            traceback.print_exc()


GRFindDelayList()
GRPrintAddrInfo()
GRFindHeader()
GRHeapMetaCorruption()
GRPrintSegv()
GRDumpInFlight()
